### PR TITLE
Isolate `Encoding` logic to be extracted and enhanced upon

### PR DESCRIPTION
**Pull request by b-n**
This will likely be the first in a series of PRs.

I've gone this particular way of extracting `Encoding` to be in line with how the other `extn/core` types are implemented.

I haven't yet got plans on how this class/type will be backed, I firstly wanted to extract and isolate it (mostly to make sure it truly is isolated), and then iterate by moving the `encoding.rb` to `mruby.rs` piece by piece.

Relates: #183

### CODE REVIEW

**Comment by b-n** [comment]
Some thoughts I'm having are around the linkage between `spinoso-string` and this particular piece of work. E.g. `spinoso-string` has [it's own `Encoding` enum](https://github.com/artichoke/artichoke/blob/trunk/spinoso-string/src/encoding.rs), and it has enough functionality to provide a number of functions that would be needed on the core `Encoding` type.

I think conceptually we could go crazy and extract a `spinoso-encoding` which is more of a database of encoding than actual implementation (that would be responsibility of spinoso-string still IMO) - but less is more, but just thinking out loud about where this could possibly go if we find the need one day.

**Comment by b-n** [comment]
@lopopolo do you have a pref to bundle this into #2579, or to keep it separate?

**Review by lopopolo — APPROVED** [review]
Looks good to me.

I do kinda like the idea of a `spinoso-encoding` crate that is a database of _all_ encodings that MRI supports/Artichoke could support. And then it'd be up to `spinoso-string` to actually build in that support (and fallible conversions from `spinoso_encoding::Encoding` to `spinoso_string::Encoding` would facilitate that).

This would also get us to a point where one real enum would be able to be used for defining all of the encoding constants in the `Encoding` class.

**Comment by lopopolo** [comment]
> @lopopolo do you have a pref to bundle this into #2579, or to keep it separate?

let's merge this separately. this is a nice refactor.

**Comment by lopopolo** [comment]
> I haven't yet got plans on how this class/type will be backed

if we add a new `MRB_TT_ENCODING` type tag to mruby, the encodings can be converted to a `u32` and stuffed in the `sym` field of the `mrb_value`, which would let us pass around encodings without dynamic allocation (and allow us to pack the encoding into the `flags` of `String` `mrb_value`s like is currently done).

File: artichoke-backend/src/extn/core/encoding/mod.rs [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+
+#[cfg(test)]
+mod tests {
+    use crate::test::prelude::*;
+
+    const SUBJECT: &str = "Encoding";
+    const FUNCTIONAL_TEST: &[u8] = include_bytes!("encoding_test.rb");
+
+    #[test]
+    fn functional() {
```
Comment (lopopolo) [inline artichoke-backend/src/extn/core/encoding/mod.rs:26]: I do love this pattern and love when new functional test suites get added. thanks for adding this.

**Comment by b-n** [comment]
> I do kinda like the idea of a spinoso-encoding crate that is a database of all encodings that MRI supports/Artichoke could support. And then it'd be up to spinoso-string to actually build in that support (and fallible conversions from spinoso_encoding::Encoding to spinoso_string::Encoding would facilitate that).

You've read my mind a bit here. I was going towards this and then spotted the Encoding being available in spinoso-striing, and thought: "Well I might aswell begin there" haha.

**Pull request merged**